### PR TITLE
[Core] Move TextProperty, ITextElement, and IPlaceholderElement to InputView

### PR DESCRIPTION
**Pull request by jcmanke**
### Description of Change ###

The Entry, Editor, and SearchBar classes are all subclasses of InputView, but each separately implements the interfaces ITextElement and IPlaceholderElement and a Text property. This PR moves these properties into InputView while maintaining ABI backwards compatibility.

### Issues Resolved ### 

- N/A

### API Changes ###

Changed:
- Moved `Text` property, `TextChanged` event, `OnTextChanged` method, `IPlaceholderElement` implementation, `ITextElement` implementation from `Entry`, `Editor`, and `SearchBar` to `InputView`

### Platforms Affected ### 

- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

N/A

### Testing Procedure ###

N/A

### PR Checklist ###

- [ ] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

**Comment by jcmanke** [comment]
@StephaneDelcroix So keep all the BindableProperty static fields like so: `public new static readonly BindableProperty TextProperty = InputView.TextProperty;` ?

**Comment by StephaneDelcroix** [comment]
@jcmanke yes, please

**Comment by jcmanke** [comment]
@StephaneDelcroix Done. I left `IsTextPredictionEnabledProperty` off of Editor because it hasn't been released yet, but I did include it for Entry.

**Comment by StephaneDelcroix** [comment]
I really have to think, and mainly test, to see if this is an ABI test for:
- the BindableProperties
- the normal properties

**Comment by jcmanke** [comment]
Now that #3248 has been merged, I'm rethinking if the Completed event should be moved to InputView, as it will sort of conflict with the SearchBar's SearchButtonPressed event.

**Comment by rmarinho** [comment]
@jcmanke can you rebase ? thanks

**Comment by jcmanke** [comment]
@rmarinho Done.

File: Xamarin.Forms.Core/InputView.cs [hunk 1]
```diff
@@ -1,41 +1,102 @@
...
+
+		public event EventHandler<TextChangedEventArgs> TextChanged;
+
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public virtual void SendCompleted() => Completed?.Invoke(this, EventArgs.Empty);
+
+		protected virtual void OnTextChanged(string oldValue, string newValue)
+		{
+			TextChanged?.Invoke(this, new TextChangedEventArgs(oldValue, newValue));
+		}
```
Comment (paymicro) [inline Xamarin.Forms.Core/InputView.cs]: Could you format this method as the previous one. Or previous format as it is.
```OnTextChanged(string oldValue, string newValue) => TextChanged?... ```

**Comment by jsuarezruiz** [comment]
@jcmanke Can you rebase and review thew conflicts?. Thanks.

**Comment by jcmanke** [comment]
@jsuarezruiz Merged. I moved the CharacterSpacing property added in #5167 out of Entry/Editor/SearchBar into InputView.

**Comment by jcmanke** [comment]
Everyone's making changes to InputView subclasses today...

**Comment by samhouts** [comment]
Sorry, @jcmanke !! Thanks for being on top of it!!

**Comment by jcmanke** [comment]
@StephaneDelcroix @jsuarezruiz What would it take to get this PR completed and included in 4.3.0? It's been sitting around for a long time, and if new stuff like CharacterSpacing gets released without this I'm starting to think it would be better to close this and start over from scratch.

**Comment by StephaneDelcroix** [comment]
I did a quick text today, and it's fine, ABI wise, to move BindableProperties, Properties and events to base class, at least on mono.

I still have to run the same test on .NET

**Comment by StephaneDelcroix** [comment]
but .NET is not as happy... and even if it doesn't fail on mono, there are some quirks

- fields are required to be there (with `new` keyword)
- properties can be move to base class, as they're invoked with `callvirt`
- events can be moved to base class, as the adder is invoked with `callvirt`

@jcmanke do you mind to review the PR based on this findings, or should I do it ?

**Comment by StephaneDelcroix** [comment]
I don't think any XamlC generated IL keeps any reference to the actual property, or the actual event, but only to accessors

**Comment by jcmanke** [comment]
@StephaneDelcroix I believe the PR as currently written matches those .NET requirements but I would appreciate another set of eyes to take a second pass.

File: Xamarin.Forms.Core/Editor.cs [hunk 2]
```diff
@@ -1,31 +1,25 @@
...
 		public static readonly BindableProperty FontFamilyProperty = FontElement.FontFamilyProperty;
 
 		public static readonly BindableProperty FontSizeProperty = FontElement.FontSizeProperty;
 
 		public static readonly BindableProperty FontAttributesProperty = FontElement.FontAttributesProperty;
 
-		public static readonly BindableProperty TextColorProperty = TextElement.TextColorProperty;
+		public new static readonly BindableProperty TextColorProperty = InputView.TextColorProperty;
 
-		public static readonly BindableProperty CharacterSpacingProperty = TextElement.CharacterSpacingProperty;
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Editor.cs]: bad merge ?

File: Xamarin.Forms.Core/Editor.cs [hunk 3]
```diff
@@ -1,31 +1,25 @@
...
+		public new static readonly BindableProperty PlaceholderProperty = InputView.PlaceholderProperty;
 
-		public static readonly BindableProperty PlaceholderProperty = PlaceholderElement.PlaceholderProperty;
-
-		public static readonly BindableProperty PlaceholderColorProperty = PlaceholderElement.PlaceholderColorProperty;
-
-		public static readonly BindableProperty IsTextPredictionEnabledProperty = BindableProperty.Create(nameof(IsTextPredictionEnabled), typeof(bool), typeof(Editor), true, BindingMode.Default);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Editor.cs]: bad merge ?

File: Xamarin.Forms.Core/Entry.cs [hunk 4]
```diff
@@ -7,25 +7,23 @@
...
 
-		public static readonly BindableProperty IsPasswordProperty = BindableProperty.Create("IsPassword", typeof(bool), typeof(Entry), default(bool));
+		public static readonly BindableProperty IsPasswordProperty = BindableProperty.Create(nameof(IsPassword), typeof(bool), typeof(Entry), default(bool));
 
-		public static readonly BindableProperty TextProperty = BindableProperty.Create("Text", typeof(string), typeof(Entry), null, BindingMode.TwoWay, propertyChanged: OnTextChanged);
+		public new static readonly BindableProperty TextProperty = InputView.TextProperty;
 
-		public static readonly BindableProperty TextColorProperty = TextElement.TextColorProperty;
-
-		public static readonly BindableProperty CharacterSpacingProperty = TextElement.CharacterSpacingProperty;
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Entry.cs]: ?

File: Xamarin.Forms.Core/SearchBar.cs [hunk 5]
```diff
@@ -38,9 +31,7 @@ public class SearchBar : InputView, IFontElement, IPlaceholderElement, ITextElem
 
 		public static readonly BindableProperty VerticalTextAlignmentProperty = TextAlignmentElement.VerticalTextAlignmentProperty;
 
-		public static readonly BindableProperty TextColorProperty = TextElement.TextColorProperty;
-
-		public static readonly BindableProperty CharacterSpacingProperty = TextElement.CharacterSpacingProperty;
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/SearchBar.cs]: ?

**Comment by StephaneDelcroix** [comment]
@jcmanke some bp overrides are missing, probably a bad merge at some point. could you fix that, and rebase to current master ? thanks

**Comment by jcmanke** [comment]
@StephaneDelcroix As I mentioned in a comment earlier, I intentionally removed CharacterSpacing from Entry/Editor/SearchBar and added it to InputView, as it's a new property that hasn't been released yet.

For the other ones you pointed out it's probably a similar thing except those were added and released between this PR being created and now.

**Comment by jcmanke** [comment]
@StephaneDelcroix Rebased on master and cleaned up a bit.

**Comment by StephaneDelcroix** [comment]
> I intentionally removed CharacterSpacing from Entry/Editor/SearchBar and added it to InputView, as it's a new property that hasn't been released yet.

CharacterSpacing is part of 4.3, but I see that you moved it up and added the new BP, so all is good

**Comment by rssllgrrtt** [comment]
I came here to talk about this. Only to find it seems to of already been resolved. So I'll just drop this comment and wait for 4.4.0.

:)

**Pull request merged**